### PR TITLE
WT-8604 Pass visible all flag to tree walk

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Review by keithbostic — COMMENTED** [review]
@quchenhao, can you explain what you're thinking here, is there any discussion of this change?

As I understand it, the previous change inside of `__txn_visible_id()` made the statement "when checking if a txn can see a TXN/TS, if we run out of  checks we can make without a snapshot, and the caller doesn't have a snapshot, switch to a global visibility check".

I agree that's not good.

I think this changes that statement to be "when checking if a fast-deleted page is visible to the caller, the fact that the caller has a snapshot means the caller wants to check if the TXN/TS is visible to the caller, otherwise the caller wants to check global visibility."

Why do we know that to be a fact? Can't a caller with a snapshot want to know if a fast-delete is globally visible?

This change is in code I don't understand very well, but it feels fragile to me.

**Comment by keithbostic** [comment]
@quchenhao, I tried to look at the `format` core dump in the evergreen failure, but I couldn't get any symbols for it. If you are able to get symbols for it, please let me know how you did it. -- Thanks!

**Comment by quchenhao** [comment]
> @quchenhao, can you explain what you're thinking here, is there any discussion of this change?
> 
> As I understand it, the previous change inside of `__txn_visible_id()` made the statement "when checking if a txn can see a TXN/TS, if we run out of checks we can make without a snapshot, and the caller doesn't have a snapshot, switch to a global visibility check".
> 
> I agree that's not good.
> 
> I think this changes that statement to be "when checking if a fast-deleted page is visible to the caller, the fact that the caller has a snapshot means the caller wants to check if the TXN/TS is visible to the caller, otherwise the caller wants to check global visibility."
> 
> Why do we know that to be a fact? Can't a caller with a snapshot want to know if a fast-delete is globally visible?
> 
> This change is in code I don't understand very well, but it feels fragile to me.

I only changed every place that pass visible all as false to the function call. In these calls, the original intention is to check with a snapshot (in other words, they assume we should have a snapshot, which may be wrong now). Everything else calls this function with visible all as true is not affected. If some function even with snapshot also want to just check global visibility, it should just pass true.

**Comment by quchenhao** [comment]
> @quchenhao, I tried to look at the `format` core dump in the evergreen failure, but I couldn't get any symbols for it. If you are able to get symbols for it, please let me know how you did it. -- Thanks!

I run `/opt/mongodbtoolchain/v3/bin/gdb t dump_t.40726.core` and it is working.

**Comment by keithbostic** [comment]
> I only changed every place that pass visible all as false to the function call. In these calls, the original intention is to check with a snapshot (in other words, they assume we should have a snapshot, which may be wrong now). Everything else calls this function with visible all as true is not affected. If some function even with snapshot also wants to just check global visibility, it should just pass true.

There's no way for a caller of `__tree_walk_internal()` to "just pass true" -- that's the problem here, this call is being made at an intermediate level of the library, the tree walk code.

File: src/btree/bt_curprev.c [hunk 1]
```diff
@@ -766,6 +766,9 @@ __wt_btcur_prev(WT_CURSOR_BTREE *cbt, bool truncating)
         if (F_ISSET(cbt, WT_CBT_READ_ONCE))
             LF_SET(WT_READ_WONT_NEED);
 
+        if (!F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT))
+            LF_SET(WT_READ_VISIBLE_ALL);
+
```
Comment (keithbostic) [inline src/btree/bt_curprev.c:771]: Should this assert we have a snapshot? That is, are we planning to fully support code paths where we don't have a snapshot, or are we planning to turn them on only by demand?

I'm slightly in favor of the latter, that is, asserting we have a snapshot until we know there's a code path where we legitimately don't have a snapshot. I could be talked out of that, though.

Edit -- from later changes, it looks to me like this is a path where you did identify a caller that doesn't have a snapshot transaction? In which case, please ignore my comment.

File: src/btree/bt_walk.c [hunk 2]
```diff
@@ -424,10 +425,14 @@ __tree_walk_internal(WT_SESSION_IMPL *session, WT_REF **refp, uint64_t *walkcntp
                 if (LF_ISSET(WT_READ_NO_WAIT) && current_state != WT_REF_MEM)
                     break;
             } else if (current_state == WT_REF_DELETED) {
+                /* Ensure we have a snapshot to check visibility or we only check global visibility.
+                 */
+                WT_ASSERT(session,
+                  LF_ISSET(WT_READ_VISIBLE_ALL) || F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT));
```
Comment (keithbostic) [inline src/btree/bt_walk.c:431]: Let's assert this at the top of the function so we don't do it on every deleted ref?

File: src/txn/txn_rollback_to_stable.c [hunk 3]
```diff
@@ -1234,13 +1234,16 @@ __rollback_abort_fast_truncate(
...
+__wt_rts_page_skip(
+  WT_SESSION_IMPL *session, WT_REF *ref, void *context, bool visible_all, bool *skipp)
 {
     wt_timestamp_t rollback_timestamp;
 
     rollback_timestamp = *(wt_timestamp_t *)(context);
     *skipp = false; /* Default to reading */
 
+    WT_UNUSED(visible_all);
+
```
Comment (keithbostic) [inline src/txn/txn_rollback_to_stable.c:1246]: This should go at the top of the function?

**Review by keithbostic — APPROVED** [review]
lgtm

File: src/btree/bt_curprev.c (hunk 1 above)
Comment (quchenhao) [reply src/btree/bt_curprev.c:771]: rollback to stable and stats cursor are the ones I found.

File: src/btree/bt_walk.c (hunk 2 above)
Comment (quchenhao) [reply src/btree/bt_walk.c:431]: I tried that and it hits too many errors. Let me try again.

File: src/txn/txn_rollback_to_stable.c (hunk 3 above)
Comment (quchenhao) [reply src/txn/txn_rollback_to_stable.c:1246]: I will hit an error if I do that.

Comment (keithbostic) [reply src/txn/txn_rollback_to_stable.c:1246]: OK, fbm.

**Comment by quchenhao** [comment]
@hbokhari @kommiharibabu Can you guys have a look?

File: src/include/txn_inline.h [hunk 4]
```diff
@@ -724,9 +724,8 @@ __txn_visible_id(WT_SESSION_IMPL *session, uint64_t id)
     if (txn->isolation == WT_ISO_READ_UNCOMMITTED)
         return (true);
 
-    /* Otherwise, we need a snapshot; if we don't have one, assume a test for global visibility. */
-    if (!F_ISSET(txn, WT_TXN_HAS_SNAPSHOT))
-        return (__txn_visible_all_id(session, id));
+    /* Otherwise, we should be called with a snapshot. */
+    WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) || session->dhandle->checkpoint != NULL);
```
Comment (keithbostic) [inline src/include/txn_inline.h:728]: Do you know why this path checks for a checkpoint handle?

File: src/reconcile/rec_child.c [hunk 5]
```diff
@@ -94,7 +96,7 @@ __rec_child_deleted(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *ref, WT_C
      * Otherwise, we have to write a proxy record. If the delete state is not ready, then delete is
      * not visible as it is in prepared state.
      */
-    if (!__wt_page_del_active(session, ref, false))
+    if (!__wt_page_del_active(session, ref, !F_ISSET(txn, WT_TXN_HAS_SNAPSHOT)))
```
Comment (keithbostic) [inline src/reconcile/rec_child.c:99]: Is this still necessary? Didn't @kommiharibabu change it recently so all eviction threads have snapshots?

**Review by keithbostic — APPROVED** [review]
lgtm!

**Comment by keithbostic** [comment]
@quchenhao, a couple of questions, I'm good to go.

File: src/reconcile/rec_child.c (hunk 5 above)
Comment (quchenhao) [reply src/reconcile/rec_child.c:99]: Yes, that may not be required.

File: src/include/txn_inline.h (hunk 4 above)
Comment (quchenhao) [reply src/include/txn_inline.h:728]: No, I don't. Let me remove this check and see how it goes.

Comment (keithbostic) [reply src/include/txn_inline.h:728]: The reason I ask -- all of the plans for getting checkpoint cursors back involve the checkpoint cursor having an appropriate snapshot. :)

Comment (quchenhao) [reply src/include/txn_inline.h:728]: It hits:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007f9f856068b1 in __GI_abort () at abort.c:79
#2  0x00007f9f843c796c in __wt_abort (session=0x5556a4ccf930) at ../src/os_common/os_abort.c:30
#3  0x00007f9f8424dd46 in __txn_visible_id (session=0x5556a4ccf930, id=5) at ../src/include/txn_inline.h:728
#4  0x00007f9f8424ddeb in __wt_txn_visible (session=0x5556a4ccf930, id=5, timestamp=50) at ../src/include/txn_inline.h:747
#5  0x00007f9f8424dba9 in __wt_txn_tw_start_visible (session=0x5556a4ccf930, tw=0x7ffe390f9da0)
    at ../src/include/txn_inline.h:637
#6  0x00007f9f8424e67d in __wt_txn_read (session=0x5556a4ccf930, cbt=0x5556a5143150, key=0x0, recno=1, upd=0x0)
    at ../src/include/txn_inline.h:1054
#7  0x00007f9f842526d1 in __cursor_var_next (cbt=0x5556a5143150, newpage=true, restart=false, skippedp=0x7ffe390f9f10)
    at ../src/btree/bt_curnext.c:312
#8  0x00007f9f84253ed1 in __wt_btcur_next_prefix (cbt=0x5556a5143150, prefix=0x0, truncating=false)
    at ../src/btree/bt_curnext.c:775
#9  0x00007f9f8425460e in __wt_btcur_next (cbt=0x5556a5143150, truncating=false) at ../src/btree/bt_curnext.c:893
#10 0x00007f9f84320121 in __curfile_next (cursor=0x5556a5143150) at ../src/cursor/cur_file.c:93
#11 0x00007f9f8473633f in _wrap_Cursor_next ()
```
I don't understand why we still can use the transaction if we don't have one.

Comment (keithbostic) [reply src/include/txn_inline.h:728]: Not terribly surprising.  Thank you for checking!

Comment (quchenhao) [reply src/include/txn_inline.h:728]: The failure does come from the checkpoint cursor.

Comment (sauclovian-wt) [reply src/include/txn_inline.h:728]: The existing checkpoint cursor code is slightly weird in this regard; effectively it acts like it has a snapshot (that comes from the checkpoint) but there isn't actually one there.

**Review by hbokhari — APPROVED** [review]
Nice work. LGTM!
